### PR TITLE
Display Key in UploadKeyActivity Issue #506

### LINKED ISSUE #506: Display key in UploadKeyActivity
Display key in UploadKeyActivity like it is done in CertifyKeyActivity.

To prevent duplicated code put this view and the loader in a fragment

### PR DESCRIPTION
**Pull request by 007durgesh219**
screenshot_2016-03-07-17-36-11
Fix #506

### CODE REVIEW

**Comment by Valodim** [comment]
Huh. Do the checkmarks actually do anything here?

**Comment by 007durgesh219** [comment]
No, It's useless. We can hide it. I thought of hiding. Just wanted to discuss before doing that. I will hide it.

**Comment by 007durgesh219** [comment]
I've updated the PR. Please review it.

screenshot_20160313-002038

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/CertifyKeyActivity.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 public class CertifyKeyActivity extends BaseActivity {
 
     public static final String EXTRA_RESULT = "operation_result";
-    public static final String EXTRA_KEY_IDS = "extra_key_ids";
+    public static final String EXTRA_KEY_IDS = MultiUserIdsFragment.EXTRA_KEY_IDS ;
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/CertifyKeyActivity.java]: we don't do this anywhere else, please leave the constants individual

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/adapter/MultiUserIdsAdapter.java [hunk 2]
```diff
@@ -138,6 +145,11 @@ public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
             }
         });
         vCheckBox.setClickable(false);
+        if (checkboxVisibility) {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/adapter/MultiUserIdsAdapter.java]: nicer with ?: operator

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/UploadKeyActivity.java [hunk 3]
```diff
@@ -89,15 +90,6 @@ public void onClick(View v) {
             return;
         }
 
-        try {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/UploadKeyActivity.java]: why was this removed?

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/UploadKeyActivity.java [hunk 4]
```diff
@@ -53,8 +50,8 @@
 
     // CryptoOperationHelper.Callback vars
     private String mKeyserver;
-    private long mMasterKeyId;
     private CryptoOperationHelper<UploadKeyringParcel, UploadResult> mUploadOpHelper;
+    private MultiUserIdsFragment mMultiUserIdsFragment;
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/UploadKeyActivity.java]: can be a local var

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/CertifyKeyFragment.java [hunk 5]
```diff
@@ -62,52 +62,19 @@
...
-        }
-
-        ArrayList<Boolean> checkedStates;
-        if (savedInstanceState != null) {
-            checkedStates = (ArrayList<Boolean>) savedInstanceState.getSerializable(ARG_CHECK_STATES);
-            // key spinner and the checkbox keep their own state
-        } else {
-            checkedStates = null;
-
+        if (savedInstanceState == null){
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/CertifyKeyFragment.java]: code style

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+import org.sufficientlysecure.keychain.provider.KeychainContract;
+import org.sufficientlysecure.keychain.provider.KeychainDatabase;
+import org.sufficientlysecure.keychain.service.CertifyActionsParcel;
+import org.sufficientlysecure.keychain.ui.adapter.MultiUserIdsAdapter;
+import org.sufficientlysecure.keychain.util.Log;
+
+import java.util.ArrayList;
+
+/**
+ * Created by durgeshchoudhary on 07/03/16.
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: pointless comment, please remove

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    public void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+
+        ArrayList<Boolean> states = mUserIdsAdapter.getCheckStates();
+        // no proper parceling method available :(
+        outState.putSerializable(ARG_CHECK_STATES, states);
+    }
+
+    public ArrayList<CertifyActionsParcel.CertifyAction> getSelectedCertifyActions() {
+        return mUserIdsAdapter.getSelectedCertifyActions();
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: throw AssertionError if checkboxVisibility is false

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+import java.util.ArrayList;
+
+/**
+ * Created by durgeshchoudhary on 07/03/16.
+ */
+public class MultiUserIdsFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor>{
+    public static final String ARG_CHECK_STATES = "check_states";
+    public static final String EXTRA_KEY_IDS = "extra_key_ids";
+    // Indicating whether checkbox should be visible or not (Default visible)
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: pointless comment: either the comment is unneeded, or the variable name isn't descriptive enough

**Comment by Valodim** [comment]
nicely done with the fragment extraction :+1:

(sorry for the late review)

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java (hunk 6 above)
Comment (007durgesh219) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: Those are automatically inserted by AndroidStudio  Anyway I will remove them. Thanks for pointing out

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/UploadKeyActivity.java (hunk 3 above)
Comment (007durgesh219) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/UploadKeyActivity.java]: Because now there is no need of getting masterKeyId from DataUri. As I am sending masterKeyId in Intent itself while calling this activity itself.  So, this try-cache is moved there.  Please see ViewKeyAdvShareFragment.java where I am calculating masterKeyId from dataUri and sending to this activity with Intent.

**Comment by 007durgesh219** [comment]
I've updated PR with suggested changes. Please review it again

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java [hunk 9]
```diff
@@ -455,8 +455,17 @@ protected void onPostExecute(Bitmap qrCode) {
     }
 
     private void uploadToKeyserver() {
+        long keyId = 0;
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java]: this error seems like it's not handled correctly

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    public void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+
+        ArrayList<Boolean> states = mUserIdsAdapter.getCheckStates();
+        // no proper parceling method available :(
+        outState.putSerializable(ARG_CHECK_STATES, states);
+    }
+
+    public ArrayList<CertifyActionsParcel.CertifyAction> getSelectedCertifyActions() {
+        assert(checkboxVisibility);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: the assert statement is optimized away in release builds, but we do want a crash report if this happens for some reason. please use `if (...) { throw new AssertionError("description"); }`

**Comment by Valodim** [comment]
the "Identities for" headline is not correctly aligned, there is a little too much space to the left (can also be seen in your screenshot). just tested it and looks good otherwise! :+1:

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java (hunk 9 above)
Comment (007durgesh219) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java]: I think adding return in catch block will do. Is that what you mean??

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java (hunk 10 above)
Comment (007durgesh219) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: Should I also include throws AssertionError to explicitly tell that this method might through an exception So, that it can be handled by the caller.

Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/MultiUserIdsFragment.java]: The point of [errors](https://docs.oracle.com/javase/7/docs/api/java/lang/Error.html) is that they are never caught by the caller. This is an illegal state caused by an error in programming that should be fixed, so we want the app to crash if it happens.

**Comment by 007durgesh219** [comment]
I've updated the PR. Now its aligned.  Please Review it again. 
screenshot_20160331-165237

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java (hunk 9 above)
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java]: Correct. In this case, the keyId variable doesn't have to be initialized - if the compiler complains about that, it means the variable might be used when there's a 0 in there, which is a sign that something is awry.

Apart from that, we use snackbars instead of toasts

Comment (007durgesh219) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ViewKeyAdvShareFragment.java]: I've updated it with suggested changes

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/adapter/MultiUserIdsAdapter.java [hunk 11]
```diff
@@ -46,6 +47,11 @@ public MultiUserIdsAdapter(Context context, Cursor c, int flags, ArrayList<Boole
         mCheckStates = preselectStates == null ? new ArrayList<Boolean>() : preselectStates;
     }
 
+    public MultiUserIdsAdapter(Context context, Cursor c, int flags, ArrayList<Boolean> preselectStates, boolean checkboxVisibility) {
+        this(context,c,flags,preselectStates);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/adapter/MultiUserIdsAdapter.java]: code style: spaces

**Comment by Valodim** [comment]
Cool, looks good now! Thanks for your work :+1:
